lib/lightstep/tracer/span.rb

```diff
@@ -0,0 +1,132 @@
+# TODO(ngauthier@gmail.com) Separate Span and SpanContext under a getter according
+# to the spec. Baggage moves to span context.
+module LightStep
+  # Span represents an OpenTracer Span
+  #
+  # See http://www.opentracing.io for more information.
+  class Span
+    # UnsupportedValueTypeError is raised when a tag value is not a supported type
+    class UnsupportedValueTypeError < LightStep::Error; end
+
+    # The guid of the span
+    attr_reader :guid
+    # Tags on the span
+    attr_reader :tags
+    # The baggage attached to this span
+    attr_reader :baggage
+    # The {Tracer} that created this span
+    attr_reader :tracer
+
+    # The guid of the current trace
+    attr_accessor :trace_guid
+    # The operation name
+    attr_accessor :operation_name
+    # Start time of the span in microseconds
+    attr_accessor :start_micros
+    # End time of the span in microseconds
+    attr_accessor :end_micros
+
+    # Creates a new {Span}
+    #
+    # @param tracer [Tracer] the tracer that created this span
+    # @param operation_name [String] the operation name of this span
+    # @param child_of_guid [String] the guid of the span this span is a child of
+    # @param trace_guid [String] the guid of this span's trace
+    # @param start_micros [Numeric] start time of the span in microseconds
+    # @return [Span] a new Span
+    def initialize(
+      tracer:,
+      operation_name:,
+      child_of_guid: nil,
+      trace_guid:,
+      start_micros:,
+      tags: nil
+    )
+      @tags = Hash(tags)
+      @baggage = {}
+
+      @tracer = tracer
+      @guid = tracer.generate_guid
+      self.operation_name = operation_name
+      self.start_micros = start_micros
+      self.trace_guid = trace_guid
+      set_tag(:parent_span_guid, child_of_guid) if !child_of_guid.nil?
+    end
+
+    # Set a tag value on this span
+    # @param key [String] the key of the tag
+    # @param value [String, Numeric, Boolean] the value of the tag
+    def set_tag(key, value)
+      case value
+      when String, Fixnum, TrueClass, FalseClass
+        @tags[key] = value
+      else
+       raise UnsupportedValueTypeError,
```

Comment:
Right now there's no error logging at all in the gem. When I started working on it I found a lot of "TODO report errors here". Since I don't know how to log internal errors because I don't know the server's api for that I couldn't do this. But I agree that they should be logged to LightStep and not raised, but I didn't want to drop them silently.